### ISSUE TITLE
ClientConnectionPool: Error after server sends "Connection: closed" header

### ISSUE THREAD

**Issue by samarai-jk** [oatpp/oatpp#443]
I ran into an problem using an ApiClient with ClientConnectionPool. 

In our case the server (NodeJs/Express) is sitting behind an AWS application load-balancer. With certain requests the load-balancer drops the "Connection: keep-alive" returned by our server and sets "Connection: closed", and closes the internal connection (which is IMO a bug in the load-balancer but that's another story).

On the client side, the oatpp-ApiClient happily reuses that connection if the next request is sooner than the pool-timeout which leads to "Server returned status 0"-error in debug mode and a "Network error X" in release mode.

I was able to solve the problem by checking the server response in `HttpRequestExecutor::executeOnce` and invalidating the connection:

```cpp
  if (auto con_hdr = result.headers.get("Connection"))
  {
	  auto val = con_hdr->std_str();
	  if (val == "close")
		  invalidateConnection(connectionHandle);
  }
```

Now the dropped pool-connections are re-established and we don't get errors. The question is of course... is that the appropriate way to do this? Might this have any unwanted side-effects?

Thanks, Jiri

**Comment by lganzzzo** [comment]
Hello @samarai-jk ,

Thank you for the issue description!

Yes, it's a proper way to do it. 
Also, here are better code examples:

### N1 - the same as yours 

```cpp
if (result.headers.get("Connection") == "close") {
  invalidateConnection(connectionHandle);
}
```

### N2 - Case insensitive comparison

```cpp
auto con_hdr = result.headers.getAsMemoryLabel<oatpp::data::share::StringKeyLabelCI>("Connection");
if(con_hdr == "close") {
  invalidateConnection(connectionHandle);
}
```

The `N2` is better because some servers can respond with `Connection: close` and others with `Connection: Close`.

Please feel free to send a PR with this improvement.

Regards,
Leonid

**Comment by samarai-jk** [comment]
Thanks for the reply @lganzzzo! I can confirm your solution `N2` works on Windows 10/64 and have sent a PR #444.

Regards,
Jiri

**Comment by samarai-jk** [comment]
It turns out this seems to have side effects... bad ones. I am not 100% sure about what is happening but depending on response size (at least a couple of kb) there seems to be a timing problem. Is it possible that the response is not fully read at the time that `HttpRequestExecutor::executeOnce` is processed? 

When I call `response->transferBodyToStream()` of a connection that was invalidated in `HttpRequestExecutor` the success seems to depend on the body size. If the body is smaller than a couple of kb (4?) the file is written fully. If the body is larger, the file is cut just below 4kB but no exception is thrown.

How can we invalidate the connection only once it is sure the body has been processed? I tried to call `api_client->invalidateConnection(api_client->getConnection())` after the body is processed but that doesn't seem to be doing anything.

Any input?

Thanks!
Jiri

**Comment by lganzzzo** [comment]
Hey @samarai-jk ,

> It turns out this seems to have side effects... bad ones. I am not 100% sure about what is happening but depending on response size (at least a couple of kb) there seems to be a timing problem.

That's because we invalidate connection (closing) before we read the body.

The correct workflow should be like that:

```cpp
{
  auto connection = apiClient->getConnection(); //<-- get connection from pool

  auto response = apiClient->myAPICall(<params>, connection); //<-- use specific connection for API call

  // TODO read body

  if(response->getHeader("Connection") == "close") {
    apiClient->invalidateConnection(connection);
  }

} // connection is returned to the pool at the end of a block and since it's invalid - it gets dropped from the pool.
```

> How can we invalidate the connection only once it is sure the body has been processed? I tried to call api_client->invalidateConnection(api_client->getConnection()) after the body is processed but that doesn't seem to be doing anything.

```cpp
invalidateConnection(api_client->getConnection()) //<-- in this case, you are invalidating the newly created connection.
```

---

Also, we should think about how to invalidate connections automatically if the `connection: close` header was present...

Regards,
Leonid.

**Comment by samarai-jk** [comment]
It works! Thank you very much @lganzzzo ! This took me the better part of almost 2 weeks to solve (2 weeks and I still don't know what exactly AWS is doing).

I had to add your workflow to each request, manually checking for invalid connections, and now everything seems stable (fingers crossed, until AWS-ALB finds a new way to mess with the traffic). The trick was to understand that (A) apiClient->getConnection() creates a new connection and (B) you can actually specify the connection for each API requests.

I added a simple method to our ApiClient/DTO (using `boost::iequals` for case-insensitive string comparison) which is called after each client request:

```cpp
void CheckClose(Response& res, std::shared_ptr<RequestExecutor::ConnectionHandle> connection)
{
	if (auto con_hdr = res.getHeader("Connection"))
		if (boost::iequals(con_hdr->std_str(), "close"))
			invalidateConnection(connection);
}
```
My PR #444 should be deleted as it actually introduces a new problem.

As for how to invalidate the connections automatically, let me know if you have an idea how it could be done, I wasn't able to come up with a solution. Happy to contribute though if I can. The problem should probably be at least documented somehow, sure an AWS load-balancer is not the only scenario where that could happen.

Thanks again,
Jiri

**Comment by lganzzzo** [comment]
> As for how to invalidate the connections automatically, let me know if you have an idea how it could be done, I wasn't able to come up with a solution. Happy to contribute though if I can. The problem should probably be at least documented somehow, sure an AWS load-balancer is not the only scenario where that could happen.

I think we can take advantage of [HttpRequestExecutor::HttpConnectionHandle](https://github.com/oatpp/oatpp/blob/master/src/oatpp/web/client/HttpRequestExecutor.hpp#L53) and its' destructor. 
For example - if `connection: close` header was present in the response - we can set `ConnectionHandler` to invalidate the connection on its destruction.

We can try to sort it out by the `1.3.0` release.

**Comment by lganzzzo** [comment]
Since #456 no extra code is needed.
Now connection will be automatically invalidated on release when `Connection: close` header is present.

```cpp
{
  auto response = apiClient->myAPICall(<params>);
  
  // TODO read body

} //<-- connection automatically invalidated if `connection: close` header was present
```

@samarai-jk  please verify

**Comment by samarai-jk** [comment]
@lganzzzo Thanks for the update! Will check hopefully later today.

Jiri

**Comment by samarai-jk** [comment]
@lganzzzo I fetched your latest commits and it seems to work! Didn't have much time to test properly but I can see that unexpectedly closed connections are invalidated and not re-used. Very cool! Will work with this version for a while to get a bit more testing done but so far great 👍🙏

Do you have a pay-me-a-beer button somewhere?

Jiri

**Comment by lganzzzo** [comment]
Hey @samarai-jk ,

Thanks for verifying the fix!
Please let me know if there are any other issues.

> Do you have a pay-me-a-beer button somewhere?

Sure,- you can support oatpp here - https://github.com/sponsors/oatpp

Best Regards,
Leonid
